game/world/triggers/cscamera.cpp

```diff
@@ -1,13 +1,307 @@
+  }
+
+void CsCamera::tick(uint64_t dt) {
+  time += float(dt)/1000.f;
+
+  if(time>duration+delay) {
+    clear();
+    return;
+    }
+
+  if(time>duration)
+    return;
+
+  Vec3 cPos,cSpin;
+  if(posSpline.size()==1)
+    cPos = posSpline[0].c[3];
+  else if(posSpline.size()>1)
+    cPos = position(posSpline);
+
+  auto camera = Gothic::inst().camera();
+  if(targetSpline.size()==0)
+    cSpin = cPos - camera->destPosition();
+  else if(targetSpline.size()==1)
+    cSpin = targetSpline[0].c[3] - cPos;
+  else if(targetSpline.size()>1)
+    cSpin = position(targetSpline) - cPos;
+
+  camera->setPosition(cPos);
+  camera->setSpin(spin(cSpin));
+  }
+
+void CsCamera::init(const phoenix::vobs::cs_camera& cam) {
+  for(auto spl : {&posSpline,&targetSpline}) {
+    uint32_t size = spl->size();
+    float    len  = 0;
+    for(uint32_t i=0;i+1<size;++i) {
+      uint32_t k = i;
+      if(spl==&targetSpline)
+        k += uint32_t(cam.position_count);
+      auto& f   = cam.frames[k];
+      auto& kF  = spl->keyframe[i];
+      float a1  = (1+f->cam_bias) * (1-f->tension) * (1+f->continuity) * 0.5f;
+      float a2  = (1-f->cam_bias) * (1-f->tension) * (1-f->continuity) * 0.5f;
+      float a3  = (1+f->cam_bias) * (1-f->tension) * (1-f->continuity) * 0.5f;
+      float a4  = (1-f->cam_bias) * (1-f->tension) * (1+f->continuity) * 0.5f;
+      Vec3  p0  = i==0 ? kF.c[3] : spl->keyframe[i-1].c[3];
+      Vec3  p1  = kF.c[3];
+      Vec3  p2  = spl->keyframe[i+1].c[3];
+      Vec3  p3  = i+2==size ? kF.c[3] : spl->keyframe[i+2].c[3];
+      Vec3  dd  = (p1-p0)*a1 + (p2-p1)*a2;
+      Vec3  sd  = (p2-p1)*a3 + (p3-p2)*a4;
+      kF.c[0] =  2 * p1 - 2*p2 +   dd + sd;
+      kF.c[1] = -3 * p1 + 3*p2 - 2*dd - sd;
+      kF.c[2] = std::move(dd);
+      len    += kF.arcLength();
+      }
+
+    if(size<2)
+      continue;
+    assert(spl->keyframe.front().time==0);
+    assert(spl->keyframe.back().time==duration);
+    const float    vSlow  = 0;
+    const float    vMean  = len/duration;
+          uint32_t start  = spl==&posSpline ? 0 : uint32_t(cam.position_count);
+          uint32_t end    = spl==&posSpline ? uint32_t(cam.position_count-1) : uint32_t(cam.frames.size()-1);
+          auto     mType0 = cam.frames[start]->motion_type;
+          auto     mType1 = cam.frames[end]->motion_type;
+          float    d0     = mType0==phoenix::camera_motion::slow ? vSlow : vMean * duration;
+          float    d1     = mType1==phoenix::camera_motion::slow ? vSlow : vMean * duration;
+
+    spl->c[0] = - 2*len +     d0 + d1;
+    spl->c[1] = + 3*len - 2*d0 - d1;
+    spl->c[2] = d0;
+    }
+  }
+
+ float CsCamera::KeyFrame::arcLength(float t0,float t1) {
+  // based on https://www.gnu.org/software/gsl/doc/html/integration.html#qag-adaptive-integration
+  auto qGK = [this](float t0, float t1,float& area, float& error) {
+    auto f = [this](float t) {
+    Vec3 b = c[0]*t*t*3.f + c[1]*t*2.f + c[2];
+    return b.length();
+    };
+
+    static const float x[7] = {
```

Comment:
The numbers are weights and arguments for a quadrature formula for numeric integration. The code is taken from a library but in a condensed form. If we use only scale the spline arguments by time this can be removed (see above).